enterprise/cmd/executor/internal/worker/handler.go

```diff
@@ -143,13 +144,10 @@ func (h *handler) Handle(ctx context.Context, logger log.Logger, job types.Job)
 		}
 	}()
 
-	// TODO check that the original spec.Image isn't fully qualified so we don't inject our path
-	// and create an invalid docker name
-
 	// Invoke each docker step sequentially
 	for i, dockerStep := range job.DockerSteps {
 		image := dockerStep.Image
-		if options.DockerOptions.RegistryUrl != "" {
+		if options.DockerOptions.RegistryUrl != "" && dockertools.ParseImageString(image).Registry == "" {
```

Comment:
We probably want/need to do the following:

```suggestion
		if options.DockerOptions.RegistryUrl != "" && dockertools.ParseImageString(image).IsPublicDockerHub() {
```

```go
func (i ImageReference) IsPublicDockerHub() bool {
  return i.Registry == "" || i.Registry == "index.docker.io"
}
```

Reply:
I'm not sure about this. In theory most public containers are coming from Docker Hub but already there there's some issues:

```go
	// legacyDefaultDomain is the legacy domain for Docker Hub (which was
	// originally named "the Docker Index"). This domain is still used for
	// authentication and image search, which were part of the "v1" Docker
	// registry specification.
	//
	// This domain will continue to be supported, but there are plans to consolidate
	// legacy domains to new "canonical" domains. Once those domains are decided
	// on, we must update the normalization functions, but preserve compatibility
	// with existing installs, clients, and user configuration.
	legacyDefaultDomain = "index.docker.io"

	// defaultDomain is the default domain used for images on Docker Hub.
	// It is used to normalize "familiar" names to canonical names, for example,
	// to convert "ubuntu" to "docker.io/library/ubuntu:latest".
	//
	// Note that actual domain of Docker Hub's registry is registry-1.docker.io.
	// This domain will continue to be supported, but there are plans to consolidate
	// legacy domains to new "canonical" domains. Once those domains are decided
	// on, we must update the normalization functions, but preserve compatibility
	// with existing installs, clients, and user configuration.
	defaultDomain = "docker.io"
```
https://sourcegraph.com/github.com/distribution/distribution/-/blob/reference/normalize.go?L11-31&subtree=true

So we'd have to check both of those cases, and then we would miss public containers being pulled from `gcr.io` (while it's still up), `quay.io`, and any other public registries that may exist.

The intention of this check was that if you define a registry _at all_, we assume that you're fully-qualifying the container path and we do not modify that container name at all.

In the Cloud context (what this is designed for), if the user does not specify a registry, images are pulled from Docker Hub through Artifact Registry, but if you decided to pull something public (e.g. [distroless](https://github.com/GoogleContainerTools/distroless#docker)), we should support that as well.

Reply:
With docker mirrors, docker _only_ mirrors docker hub though, it would not mirror quay or anything else. 

That's at least for restoring the behavior we had before with docker-mirror. 

Do you think we should _change_ the behavior here and also pull from other public registries through the artifact registry? In that case, we might need to be more lenient here with the registry URL matching, correct. 

Another question that just came to mind, how do we handle authentication with private docker hub images? With registry mirrors (official ones) you get that working for free, if the mirror gets a 403 for that image, docker automatically falls back to using the upstream hub, and sends credentials for that, if configured. 
If we overwrite the images, we would get auth errors though, right? We would need to retry then, or something 🤔

Reply:
Please take another look when you get the chance. I think I resolved all the issues:
- Check for public docker registries
- Test for registries with a port
- Add tests for Gitlab registry formats

Regarding auth, private images cannot be pulled through Artifact Registry so I have to figure out some way of falling back to Docker Hub, either natively or some sort of retry in the code.